rewriter/flatten.cc

```diff
@@ -313,23 +323,40 @@ class FlattenWalk {
         }
         classDef->rhs = addClassDefMethods(ctx, std::move(classDef->rhs), classDef->loc);
         return classDef;
-    };
+    }
+
+    bool shouldMoveSend(unique_ptr<ast::Send> &send) {
+        if (send->fun == core::Names::sig()) {
+            return true;
+        }
+        if ((send->fun == core::Names::private_() || send->fun == core::Names::privateClassMethod() ||
+             send->fun == core::Names::protected_() || send->fun == core::Names::public_()) &&
+            send->args.size() == 1 && send->recv->isSelfReference()) {
+            return true;
+        }
+        return false;
+    }
 
     unique_ptr<ast::Send> preTransformSend(core::Context ctx, unique_ptr<ast::Send> send) {
+        auto &methods = curMethodSet();
+        methods.sendDepth++;
         // we might want to move sigs, so we mostly use the same logic that we use for methods. The one exception is
         // that we don't know the 'staticness level' of a sig, as it depends on the method that follows it (whether that
         // method has a `self.` or not), so we'll fill that information in later
-        if (send->fun == core::Names::sig()) {
-            curMethodSet().pushScope(computeScopeInfo(ScopeType::StaticMethodScope));
+        if (shouldMoveSend(send)) {
+            auto scopeType =
+                send->fun == core::Names::sig() ? ScopeType::StaticMethodScope : ScopeType::VisibilityModScope;
+            methods.pushScope(computeScopeInfo(scopeType));
```

Comment:
Is it important that `computeScopeInfo` doesn't handle `VisibilityModScope` explicitly?